core/arch/arm/include/sm/optee_smc.h

```diff
@@ -559,6 +559,76 @@
 /* See OPTEE_SMC_CALL_WITH_REGD_ARG above */
 #define OPTEE_SMC_FUNCID_CALL_WITH_REGD_ARG	U(19)
 
+/*
+ * Secure watchdog management service
+ *
+ * Call requests usage:
+ * [in]		a0	SMC Function ID, OPTEE_SMC_WATCHDOG
+ * [out]	a0	PSCI error code return
+ *
+ * Watchdog commands
+ * Command SMCWD_INIT : Watchdog initialization
+ * [in]		a1	Set to 0
+ * [out]	a1	The minimal timeout value in seconds supported
+ *		a2	The maximal timeout value in seconds supported
+ *
+ * Return codes:
+ * PSCI_RET_SUCCESS - Command success
+ * PSCI_RET_INTERNAL_FAILURE - Initialization failure
+ *
+ * Command SMCWD_SET_TIMEOUT : Watchdog set timeout
+ * [in]		a1	Set to 1
+ *		a2	The timeout value in seconds to set
+ *
+ * Return codes:
+ * PSCI_RET_SUCCESS - Command success
+ * PSCI_RET_INVALID_PARAMETERS - Incorrect input param
+ *
+ * Command SMCWD_ENABLE : Watchdog enable
+ * [in]		a1	Set to 2
+ *		a2	Set to 0 to stop the watchdog, 1 to enable it
+ *
+ * Return codes:
+ * PSCI_RET_SUCCESS - Command success
+ * PSCI_RET_INVALID_PARAMETERS - Incorrect input param
+ *
+ * Command SMCWD_PET : Ping the watchdog for refresh
+ * [in]		a1	Set to 3
+ *
+ * Return codes:
+ * PSCI_RET_SUCCESS - Command success
+ * PSCI_RET_DISABLED - The watchdog is not enabled
+ *
+ * Command SMCWD_GET_TIMELEFT : Get time left
+ * [in]		a1	Set to 4
+ * [out]	a1	The timeout value in seconds before watchdog expires
+ *
+ * Return codes:
+ * PSCI_RET_SUCCESS - Command success
+ * PSCI_RET_DISABLED - The watchdog is not enabled
+ * PSCI_RET_NOT_SUPPORTED - Function not supported
+ *
+ * Other commands
+ * [in]		a1	Other values
+ *
+ * Return codes:
+ * PSCI_RET_NOT_SUPPORTED - Function not supported
+ *
+ * a3-6		Not used
+ * a7		Hypervisor Client ID register
```

Comment:
The API is not that easy to read.
I would propose the below layout (without tabs in inline comments):
```
/*
 * Secure watchdog management service
 *
 * Call requests usage (for all watchdog command):
 * a0      [in]         SMC Function ID, OPTEE_SMC_WATCHDOG
 *         [out]        PSCI error code return
 * a1      [in]         Watchdog command (one of SMCWD_*)
 *         [out]        Depends on watchdog command (input a1)
 * a2      [in/out]     Depends on watchdog command (input a1)
 * a3-6    [in/out]     Not used
 * a7      [in/out]     Hypervisor Client ID register
 *
 * For watchdog command SMCWD_INIT: Watchdog initialization
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      0 (command SMCWD_INIT)
 * [out]        a0      PSCI_RET_SUCCESS - Command success
 *                      PSCI_RET_INTERNAL_FAILURE - Initialization failure
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 * [out]        a1      The minimal timeout supported value in seconds
 *                      Applicable only upon a0 is PSCI_RET_SUCCESS
 * [out]        a2      The maximal timeout supported value in seconds
 *                      Applicable only upon a0 is PSCI_RET_SUCCESS
 *
 * For watchdog command SMCWD_SET_TIMEOUT: Watchdog set timeout
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      1 (command SMCWD_SET_TIMEOUT)
 * [in]         a2      The timeout value in seconds to set
 * [out]        a0      PSCI_RET_SUCCESS - Command success
 *                      PSCI_RET_INVALID_PARAMETERS - Incorrect input param
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 *
 * For watchdog command SMCWD_ENABLE: Watchdog enable
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      2 (command SMCWD_ENABLE)
 * [in]         a2      Set to 0 to stop the watchdog, 1 to enable it
 * [out]        a0      PSCI_RET_SUCCESS - Command success
 *                      PSCI_RET_INVALID_PARAMETERS - Incorrect input param
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 *
 * For watchdog command SMCWD_PET: Ping the watchdog for refresh
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      3 (command SMCWD_PET)
 * [out]        a0      PSCI_RET_SUCCESS - Command success
 *                      PSCI_RET_DISABLED - The watchdog is not enabled
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 *
 * For watchdog command SMCWD_GET_TIMELEFT: Get time left
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      4 (command SMCWD_GET_TIMELEFT)
 * [out]        a0      PSCI_RET_SUCCESS - Command success
 *                      PSCI_RET_DISABLED - The watchdog is not enabled
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 * [out]        a1      Guaranteed time in seconds before watchdog expires
 *                      Applicable only upon a0 is PSCI_RET_SUCCESS
 *
 * For any other command value
 * [in]         a0      SMC Function ID (OPTEE_SMC_WATCHDOG)
 * [in]         a1      > 4
 * [out]        a0      PSCI_RET_INVALID_PARAMETERS - Invalid watchdog command
 *                      PSCI_RET_NOT_SUPPORTED - Function ID (a0) not supported
 *
 * By SMCCC convention:
 * PSCI_RET_SUCCESS, OPTEE_SMC_RETURN_OK and ARM_SMCCC_RET_SUCCESS
 * are equal.
 * PSCI_RET_NOT_SUPPORTED, OPTEE_SMC_RETURN_UNKNOWN_FUNCTION and
 * ARM_SMCCC_RET_NOT_SUPPORTED are equal.
 */
#define OPTEE_SMC_FUNCID_WATCHDOG		U(20)
#define OPTEE_SMC_WATCHDOG \
	OPTEE_SMC_FAST_CALL_VAL(OPTEE_SMC_FUNCID_WATCHDOG)

#define SMCWD_INIT		0
#define SMCWD_SET_TIMEOUT	1
#define SMCWD_ENABLE		2
#define SMCWD_PET		3
#define SMCWD_GET_TIMELEFT	4

Reply:
How about keeping to the established style in this file? I believe it's clear enough.

Reply:
@jenswi-linaro It was the initial proposal, after discussion with @etienne-lms I've tried to use a similar description based on PTA description that seems to be useful for such kind of "API" description. 
I'm okay to come back to previous style used in this file if it's clear enough.

Reply:
I'm sorry to change the direction. This looks small enough to make part of the normal OPTEE_SMC ABI, but then it should also follow the style of the ABI.
By the way, do you have examples of how this is used from the Linux kernel driver and U-Boot?